Wip rgw gc defer perf fixes

pritha-srivastava opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

pritha-srivastava commented:
This comment is with respect to the last commit - I noticed that radosgw-admin gc list is slower than the --include-all version. While I was investigating the reason behind this I discovered that using in memory transitioned_objects_cache in RGWGC::list() doesn't help since for every radosgw-admin command a new RGWGC object is instantiated and hence a new transitioned_objects_cache is instantiated, and for every gc list call we make 2 cls_rgw_gc_list calls and for every gc list --include call we make one cls_rgw_gc_list call. Instead of using an in memory variable, I have replaced that with the version of gc object being set to 2 when the transition to queue happens and the version is checked each time a gc list and gc list --include call is made. If the version of the gc object (queue) has already been set to 2 by RGWGC::process, then the cls_rgw_gc_list calls are skipped. If it hasn't been changed to 2, then the first gc list or gc list --include-all call will change it to 2, and subsequent gc list calls won't issue cls_rgw_gc_list calls. I can see a slight difference in performance when I delete 50 objects in my system, although I don't know what the impact will be on a larger scale.
Before the change: gc list : 0.16 seconds gc list --include-all: 0.12 seconds
After the change: gc list: 0.08 seconds gc list --include-all: 0.08 seconds

pritha-srivastava commented:
jenkins test make check

ivancich commented on src/cls/rgw_gc/cls_rgw_gc.cc:
```diff
@@ -114,13 +114,31 @@ static int cls_rgw_gc_queue_list_entries(cls_method_context_t hctx, bufferlist *
-      CLS_LOG(5, "ERROR: cls_rgw_gc_queue_list_entries(): failed to decode urgent data\n");
-      return -EINVAL;
+  std::unordered_map<string,ceph::real_time> xattr_urgent_data_map;
+  if (g_ceph_context->_conf->rgw_gc_defer_enabled) {
+    if (head.bl_urgent_data.length() > 0) {
+      auto iter_urgent_data = head.bl_urgent_data.cbegin();
+      try {
+        decode(urgent_data, iter_urgent_data);
+      } catch (ceph::buffer::error& err) {
+        CLS_LOG(5, "ERROR: cls_rgw_gc_queue_list_entries(): failed to decode urgent data\n");
+        return -EINVAL;
+      }
+    }
+    if (urgent_data.num_xattr_urgent_entries > 0) {
+      bufferlist bl;
+      ret = cls_cxx_getxattr(hctx, "cls_queue_urgent_data", &bl);
+      if (ret < 0) {
+        CLS_LOG(0, "ERROR: %s(): cls_cxx_getxattrs() returned %d", __func__, ret);
+        return ret;
+      }
+      auto iter = bl.cbegin();
+      try {
+        decode(xattr_urgent_data_map, iter);
+      } catch (ceph::buffer::error& err) {
+        CLS_LOG(1, "ERROR: cls_rgw_gc_queue_list_entries(): failed to decode xattrs urgent data map\n");
+        return -EINVAL;
+      }
```
I feel that variations of this code are repeated a number of times. It would be nice if "urgent data" were abstracted into its own class, so the implementation details don't bleed everywhere and result in duplicated code. I realize it's a big ask, so it's only a suggestion, possibly for a future PR.

ivancich commented on src/cls/rgw_gc/cls_rgw_gc.cc:
```diff
@@ -245,16 +260,42 @@ static int cls_rgw_gc_queue_remove_entries(cls_method_context_t hctx, bufferlist
-    } catch (ceph::buffer::error& err) {
-      CLS_LOG(5, "ERROR: cls_rgw_gc_queue_remove_entries(): failed to decode urgent data\n");
-      return -EINVAL;
+  std::unordered_map<string, ceph::real_time> xattr_urgent_data_map;
+  if (g_ceph_context->_conf->rgw_gc_defer_enabled) {
+    if (head.bl_urgent_data.length() > 0) {
+      auto iter_urgent_data = head.bl_urgent_data.cbegin();
+      try {
+        decode(urgent_data, iter_urgent_data);
+      } catch (ceph::buffer::error& err) {
+        CLS_LOG(5, "ERROR: cls_rgw_gc_queue_remove_entries(): failed to decode urgent data\n");
+        return -EINVAL;
+      }
+    }
+    if (urgent_data.num_xattr_urgent_entries > 0) {
+      bufferlist bl;
+      ret = cls_cxx_getxattr(hctx, "cls_queue_urgent_data", &bl);
+      if (ret < 0) {
+        CLS_LOG(0, "ERROR: %s(): cls_cxx_getxattrs() returned %d", __func__, ret);
+        return ret;
+      }
+      auto iter = bl.cbegin();
+      try {
+        decode(xattr_urgent_data_map, iter);
+      } catch (ceph::buffer::error& err) {
+        CLS_LOG(5, "ERROR: cls_rgw_gc_queue_remove_entries(): failed to decode xattrs urgent data map");
+        return -EINVAL;
+      }
```
This looks like another copy of the same code.

ivancich commented on src/cls/rgw_gc/cls_rgw_gc.cc:
```diff
@@ -408,6 +498,23 @@ static int cls_rgw_gc_queue_update_entry(cls_method_context_t hctx, bufferlist *
     return -EINVAL;
   }
 
+  std::unordered_map<string, ceph::real_time> xattr_urgent_data_map;
+  if (urgent_data.num_xattr_urgent_entries > 0) {
+    bufferlist bl;
+    ret = cls_cxx_getxattr(hctx, "cls_queue_urgent_data", &bl);
+    if (ret < 0) {
+      CLS_LOG(0, "ERROR: %s(): cls_cxx_getxattrs() returned %d", __func__, ret);
+      return ret;
+    }
+    auto iter = bl.cbegin();
+    try {
+      decode(xattr_urgent_data_map, iter);
+    } catch (ceph::buffer::error& err) {
+      CLS_LOG(5, "ERROR: cls_rgw_gc_queue_remove_entries(): failed to decode xattrs urgent data map");
+      return -EINVAL;
```
It looks like we may have another copy here too.

ivancich reviewed (commented):
I'm unclear if this is just performance fixes or if this also fixes the underlying gc defer bug too.

pritha-srivastava commented:
> I'm unclear if this is just performance fixes or if this also fixes the underlying gc defer bug too.

I asked Casey @cbodley  on the other PR if I can add perf fixes to this in addition to gc defer bug fix, and I think he agreed (he gave a thumbs up), in case we do not want to club them together, I can create a separate PR for that.

pritha-srivastava commented on src/rgw/rgw_op.cc:
```diff
@@ -2017,8 +2017,18 @@ int RGWGetObj::handle_slo_manifest(bufferlist& bl, optional_yield y)
 {
-  /* garbage collection related handling:
-   * defer_gc disabled for https://tracker.ceph.com/issues/47866 */
+  if (s->cct->_conf->rgw_gc_defer_enabled) {
```
Can you please explain your comment.

ivancich replied:
I think the modern way of getting the config option would be something like:

`s->cct->_conf.get_val<bool>("rgw_gc_defer_enabled")`

mattbenjamin replied:
this is not correct advice;  the "modern" config options are significantly slower and we have intentionally deferred to change (integral) "legacy" unless one of the caching mechanisms can be used, until a systematic improvement in config option retrieval performance is implemented--n.b., no one in the ceph community is working on one

mattbenjamin commented on src/rgw/rgw_op.cc:
```diff
@@ -2017,8 +2017,18 @@ int RGWGetObj::handle_slo_manifest(bufferlist& bl, optional_yield y)
+        ldpp_dout(this, 0) << "WARNING: could not defer gc entry for obj" << dendl;
+      }
+      gc_invalidate_time = start_time;
+      gc_invalidate_time += (s->cct->_conf->rgw_gc_obj_min_wait / 2);
```
this is not correct advice; the "modern" config options are significantly slower and we have intentionally deferred to change (integral) "legacy" unless one of the caching mechanisms can be used, until a systematic improvement in config option retrieval performance is implemented--n.b., no one in the ceph community is working on one

mattbenjamin commented on src/rgw/rgw_gc.cc:
```diff
@@ -46,6 +46,7 @@ void RGWGC::initialize(CephContext *_cct, RGWRados *_store) {
 
     //version = 0 -> not ready for transition
     //version = 1 -> marked ready for transition
+    //version = 2 -> transitioned to queue completely
```
yes, an enum class would be good

cbodley commented:
> > I'm unclear if this is just performance fixes or if this also fixes the underlying gc defer bug too.
> 
> I asked Casey @cbodley on the other PR if I can add perf fixes to this in addition to gc defer bug fix, and I think he agreed (he gave a thumbs up), in case we do not want to club them together, I can create a separate PR for that.

i think performance fixes should be backported as well, so it makes sense to me to review/test/merge them together in the same pr

pritha-srivastava commented on src/cls/rgw_gc/cls_rgw_gc.cc:
```diff
@@ -155,58 +173,42 @@ static int cls_rgw_gc_queue_list_entries(cls_method_context_t hctx, bufferlist *
+          } else {
+            //Can break out here if info.time > now
+            is_truncated = false;
+            break;
```
It is the opposite actually, if a client requests for 100 entries and we have only 98 then we set is_truncated = false, else if there are more than 100 entries then we send 100 entries and set is_truncated to true and the next marker to the 101st element, and the client sends requests for the next 100 entries. That is why is_truncated is set to false on encountering a non expired entry, so that the client doesn't send further list requests

The same do while (truncated) is there in RGWGC::process() also.

pritha-srivastava commented on src/cls/rgw_gc/cls_rgw_gc.cc:
```diff
@@ -114,13 +114,31 @@ static int cls_rgw_gc_queue_list_entries(cls_method_context_t hctx, bufferlist *
-      CLS_LOG(5, "ERROR: cls_rgw_gc_queue_list_entries(): failed to decode urgent data\n");
-      return -EINVAL;
+  std::unordered_map<string,ceph::real_time> xattr_urgent_data_map;
+  if (g_ceph_context->_conf->rgw_gc_defer_enabled) {
```
ok

pritha-srivastava commented:
> This comment is with respect to the last commit - I noticed that radosgw-admin gc list is slower than the --include-all version. While I was investigating the reason behind this I discovered that using in memory transitioned_objects_cache in RGWGC::list() doesn't help since for every radosgw-admin command a new RGWGC object is instantiated and hence a new transitioned_objects_cache is instantiated, and for every gc list call we make 2 cls_rgw_gc_list calls and for every gc list --include call we make one cls_rgw_gc_list call. Instead of using an in memory variable, I have replaced that with the version of gc object being set to 2 when the transition to queue happens and the version is checked each time a gc list and gc list --include call is made. If the version of the gc object (queue) has already been set to 2 by RGWGC::process, then the cls_rgw_gc_list calls are skipped. If it hasn't been changed to 2, then the first gc list or gc list --include-all call will change it to 2, and subsequent gc list calls won't issue cls_rgw_gc_list calls. I can see a slight difference in performance when I delete 50 objects in my system, although I don't know what the impact will be on a larger scale.
> Before the change: gc list : 0.16 seconds gc list --include-all: 0.12 seconds
> After the change: gc list: 0.08 seconds gc list --include-all: 0.08 seconds

I have removed all related commits, since testing in a larger system showed no performance improvement. And addition of another state may make upgrading to cls fifo even more difficult.

pritha-srivastava commented:
Teuthology run results here: https://pulpito.ceph.com/prsrivas-2021-01-15_10:43:52-rgw-wip-rgw-gc-defer-perf-fixes-distro-basic-smithi/, the failures are not related to this PR.
